### PR TITLE
Fix LOD selection in compatibility backend and clean up LOD code

### LINKED ISSUE #67479: Regression in mesh LOD behavior with shadows since 4.0.beta3
### Godot version

4.0 Beta 3

### System information

Windows 11, Vulkan, GTX 1050 Ti 516.94

### Issue description

A regression since Beta 2, Directional Shadow ALODs are acting up. They're more aggressive compared to the Camera, but also inconsistent within splits.

### Steps to reproduce

Move around Shadowed Objects with ALODs (directional shadows) and observe inconsistent LODs between View and Shadows, and even changing within splits.

### Minimal reproduction project

[LOD.zip](https://github.com/godotengine/godot/files/9793839/LOD.zip)

### LINKED ISSUE #73472: Orthogonal camera shadows are lower resolution than perspective projection
### Godot version

v4.0.rc2.official [d2699dc7a]

### System information

Windows 10, Vulkan 1.3.224, NVIDIA GeForce RTX3080

### Issue description

While following [a tutorial](https://youtu.be/YiE9tcoCfhE), I observed curious shadows using an orthographic camera with a directional light. It appears changing the camera projection type between perspective and orthographic dramatically reduces the shadow quality.

With an orthographic camera:
orthographic_shadows

Whereas with a perspective camera:
perspective_shadows

Turning the shadow blur on the light to a low value shows the low resolution:
orthographic_shadows_low_blur

Even pushing the shadow map resolution to the maximum (16384) in the project settings results in less-than-ideal shadows:
orthographic_shadows_max_rez

I haven't tested myself in 3.x yet, but [the tutorial](https://youtu.be/YiE9tcoCfhE) shows the orthogonal shadows [working fine](https://youtube.com/clip/Ugkx-FTJbqyibUymxNG2QNCEdZOEMJUBuh-z) on 3.3.rc8

### Steps to reproduce

See minimal reproduction project. Create a plane, camera, directional light (shadows enabled), and some shadow caster. Set the camera to orthogonal projection.

### Minimal reproduction project

[3d-squash-the-creeps-issue.zip](https://github.com/godotengine/godot/files/10762216/3d-squash-the-creeps-issue.zip)

### LINKED ISSUE #78422: Shadows broken with orthogonal camera in 3d both in editor and in game
### Godot version

v4.0.2.stable.official [7a0977ce2]

### System information

Windows 10 (Version 10.0.19045 Build 19045) - Forward+ - Intel Core i7-7700K CPU - NVIDIA GeForce RTX 3080 (driver version 536.23)

### Issue description

Orthogonal camera in editor:
Ortho

With the perspective camera:
Perspective
I would expect the shadow to be of the same quality whether in perspective or orthogonal camera

Cinematic view orthogonal camera:
CinematicPreviewOrtho
This screenshot uses a camera further away to show that the LOD used to generate the shadow break the shadow.

### Steps to reproduce

Use the orthogonal camera.

### Minimal reproduction project

[MinimalProject.zip](https://github.com/godotengine/godot/files/11782773/MinimalProject.zip)

### LINKED ISSUE #78664: Directional Shadow Pancake Size breaks Automatic LOD
### Godot version

v4.1.beta3.official [ada712e06]

### System information

Godot v4.1.beta3 - Windows 10.0.22621 - Vulkan (Forward+) - dedicated NVIDIA GeForce GTX 1050 Ti (NVIDIA; 31.0.15.3179) - Intel(R) Core(TM) i3-10105F CPU @ 3.70GHz (8 Threads)

### Issue description

Pancake Size changes Directional Shadow LOD very aggressively, not representative of the used resolution of the object in shadow map

### Steps to reproduce

Open The Project, and Play around with Pancake Size and observe Human Dummy Shadows going to hell

### Minimal reproduction project

[Pancake.zip](https://github.com/godotengine/godot/files/11859346/Pancake.zip)

### LINKED ISSUE #89394: "Generate LODs" can produce lighting bug on imported mesh (glb), camera angle dependent
### Tested versions

Godot v4.2.1.stable

### System information

Windows 10.0.22631 - Vulkan (Forward+) - dedicated AMD Radeon RX5600 (Advanced Micro Devices, Inc.; 31.0.22023.1014) - AMD Ryzen 7 3700X 8-Core Processor (16 Threads)

### Issue description

when using "Generate LODs", from some camera angles the lighting on the imported mesh suddenly changes.
I've confirmed it happens in game. seems to happen as long as the scene is lit, regardless of worldenvironment settings.
video of lighting bug, only present when importing with "Generate LODs":

https://github.com/godotengine/godot/assets/49114816/c69ffcef-eedb-43ef-b69b-eac1ebedbbeb

### Steps to reproduce

1. Import attached glb file (below)
2. Right click the file and open as New Inherited Scene
3. Move the camera to examine the ramp from various camera angles.
4. Import again with "Generate LODs" disabled
5. The bug disappears.

### Minimal reproduction project (MRP)

the mesh in question (glb file) is attached here in zip form:
[level1mesh.zip](https://github.com/godotengine/godot/files/14563048/level1mesh.zip)

### PR DESCRIPTION
**Pull request by clayjohn**
Fixes: https://github.com/godotengine/godot/issues/67479
Fixes: https://github.com/godotengine/godot/issues/89394
Fixes: https://github.com/godotengine/godot/issues/78664
Fixes: https://github.com/godotengine/godot/issues/73472
Fixes: https://github.com/godotengine/godot/issues/78422

1. Applies the fix from https://github.com/godotengine/godot/pull/92232 to the Compatibility backend
2. Removes camera_plane from the shadow_pass struct (this was unused)
3. Uses Camera transform for shadow LOD calculations so shadows match the object. This significantly reduces shadow acne and stabilizes the shadow profile. 

I have another PR coming to fix- https://github.com/godotengine/godot/issues/81995, but I separated these changes out in the meantime as these are very safe changes

CC @mrjustaguy

### CODE REVIEW

**Comment by clayjohn** [comment]
> > Uses Camera transform for shadow LOD calculations so shadows match the object. This significantly reduces shadow acne and stabilizes the shadow profile.
> 
> It might be worth checking if this fixes shadows in orthographic camera sometimes using very low-detail shadow LODs (despite camera Z far being low), while the mesh is being drawn with its full detail.

Do you have a bug report/MRP for that? I don't think this PR will help in that situation, but I can look into it

**Comment by mrjustaguy** [comment]
Pancake issue is gone
The Original issue is, mostly gone.
There is still occasionally weird LOD decisions happening with Shadows, but it's really not a big deal unlike the prior two issues that were each in their own ways highly problematic and very obvious even without dedicated attempts to break things like this

image

image

MRP: 
[Test.zip](https://github.com/godotengine/godot/files/15421758/Test.zip)

**Comment by Calinou** [comment]
> Do you have a bug report/MRP for that? I don't think this PR will help in that situation, but I can look into it

You can see it on the last GIF in this issue: https://github.com/godotengine/godot/issues/73472

**Comment by clayjohn** [comment]
> > Do you have a bug report/MRP for that? I don't think this PR will help in that situation, but I can look into it
> 
> You can see it on the last GIF in this issue: #73472
> 
> The round object has a square shadow.

Thanks! Tested that MRP and I can confirm that this fixes the issue

File: drivers/gles3/rasterizer_scene_gles3.h [hunk 1]
```diff
@@ -640,7 +640,7 @@ class RasterizerSceneGLES3 : public RendererSceneRender {
...
 	void _fill_render_list(RenderListType p_render_list, const RenderDataGLES3 *p_render_data, PassMode p_pass_mode, bool p_append = false);
 	void _render_shadows(const RenderDataGLES3 *p_render_data, const Size2i &p_viewport_size = Size2i(1, 1));
-	void _render_shadow_pass(RID p_light, RID p_shadow_atlas, int p_pass, const PagedArray<RenderGeometryInstance *> &p_instances, const Plane &p_camera_plane = Plane(), float p_lod_distance_multiplier = 0, float p_screen_mesh_lod_threshold = 0.0, RenderingMethod::RenderInfo *p_render_info = nullptr, const Size2i &p_viewport_size = Size2i(1, 1), const Transform3D &p_main_cam_transform = Transform3D());
+	void _render_shadow_pass(RID p_light, RID p_shadow_atlas, int p_pass, const PagedArray<RenderGeometryInstance *> &p_instances, float p_lod_distance_multiplier = 0, float p_screen_mesh_lod_threshold = 0.0, RenderingMethod::RenderInfo *p_render_info = nullptr, const Size2i &p_viewport_size = Size2i(1, 1), const Transform3D &p_main_cam_transform = Transform3D());
```
Comment (lawnjelly) [inline drivers/gles3/rasterizer_scene_gles3.h:643]: Just to check: Is this function ever called where `p_main_cam_transform` is not set? :thinking: (i.e. does there have to be a default)

**Review by akien-mga — APPROVED** [review]
I'm no rendering expert, but the overall code change looks sensible to me, and this would be a welcome bugfix to test in beta1.

**Review by lawnjelly — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**